polar-core/src/vm.rs

```diff
@@ -1007,41 +981,64 @@ impl PolarVirtualMachine {
+                // Check that the left is more specific than the right.
+                let left_fields: HashSet<&Symbol> = left.fields.keys().collect();
+                let right_fields: HashSet<&Symbol> = right.fields.keys().collect();
+                if !right_fields.is_subset(&left_fields) {
+                    return self.push_goal(Goal::Backtrack);
+                }
+
+                // For each field on the right, isa its value against the corresponding value on
+                // the left.
+                for (k, v) in right.fields.iter() {
+                    let left = left
+                        .fields
+                        .get(&k)
+                        .expect("left fields should be a superset of right fields")
+                        .clone();
                     self.push_goal(Goal::Isa {
-                        left: value,
-                        right: right.clone(),
-                    })?;
-                } else {
-                    self.push_goal(Goal::Unify {
-                        left: left.clone(),
-                        right: right.clone(),
-                    })?;
+                        left,
+                        right: v.clone(),
+                    })?
                 }
             }
 
-            (_, Value::RestVariable(symbol)) => {
-                if let Some(value) = self.value(&symbol).cloned() {
-                    self.push_goal(Goal::Isa {
-                        left: left.clone(),
-                        right: value,
-                    })?;
-                } else {
-                    self.push_goal(Goal::Unify {
-                        left: left.clone(),
-                        right: right.clone(),
-                    })?;
+            (_, Value::Pattern(Pattern::Dictionary(right))) => {
+                // For each field in the dict, look up the corresponding field on the instance and
+                // then isa them.
+                for (field, right_value) in right.fields.iter() {
+                    let left_value = self.kb.read().unwrap().gensym("isa_value");
+                    let call_id = self.new_call_id(&left_value);
+                    let lookup = Goal::LookupExternal {
+                        instance: left.clone(),
+                        call_id,
+                        field: right_value.clone_with_value(Value::String(field.0.clone())),
+                        check_errors: false,
+                    };
+                    let isa = Goal::Isa {
+                        left: left.clone_with_value(Value::Variable(left_value)),
+                        right: right_value.clone(),
+                    };
+                    self.append_goals(vec![lookup, isa])?;
                 }
             }
 
-            (
-                Value::ExternalInstance(left_instance),
-                Value::Pattern(Pattern::Instance(right_literal)),
-            ) => {
+            (_, Value::Pattern(Pattern::Instance(right_literal))) => {
```

Comment:
Same idea: if the RHS is an instance literal, we're always going to check types and recurse, so combine these two expressions.